rflx/specification/parser.py

```diff
@@ -1,358 +1,973 @@
+    _cache: Cache,
+    filename: Path,
+) -> Array:
+    element_identifier = qualified_type_identifier(
+        create_id(array.f_element_type, filename), identifier.parent
+    )
 
-    def __evaluate_specification(self, specification: Specification) -> None:
-        log.info("Processing %s", specification.package.identifier)
+    try:
+        element_type = [t for t in types if element_identifier == t.identifier][0]
+    except IndexError:
+        fail(
+            f'undefined element type "{element_identifier}"',
+            Subsystem.PARSER,
+            Severity.ERROR,
+            element_identifier.location,
+        )
 
-        error = RecordFluxError()
-        self.__evaluate_types(specification, error)
-        self.__evaluate_sessions(specification)
-        error.propagate()
+    return Array(identifier, element_type, node_location(array, filename))
+
+
+def create_numeric_literal(
+    expression: Expr, _filename: Path = None, _package: ID = None, location: Location = None
+) -> rexpr.Expr:
+    num = expression.text.split("#")
+    if len(num) == 1:
+        return rexpr.Number(int(num[0]), location=location)
+    if len(num) == 3:
+        base = int(num[0])
+        return rexpr.Number(int(num[1], base), base=base, location=location)
+    raise NotImplementedError(f"Invalid numeric literal: {expression.text}")
+
+
+OPERATIONS: Dict[str, Type[rexpr.BinExpr]] = {
+    "OpIn": rexpr.In,
+    "OpNotin": rexpr.NotIn,
+    "OpEq": rexpr.Equal,
+    "OpNeq": rexpr.NotEqual,
+}
+
+
+def create_binop(
+    expression: Expr, filename: Path = None, package: ID = None, _location: Location = None
+) -> rexpr.Expr:
+    if expression.f_op.kind_name in OPERATIONS:
+        return OPERATIONS[expression.f_op.kind_name](
+            create_expression(expression.f_left, filename, package),
+            create_expression(expression.f_right, filename, package),
+            location=node_location(expression, filename),
+        )
+    if expression.f_op.kind_name in BOOLEAN_OPERATIONS:
+        return BOOLEAN_OPERATIONS[expression.f_op.kind_name](
+            create_expression(expression.f_left, filename, package),
+            create_expression(expression.f_right, filename, package),
+            location=node_location(expression, filename),
+        )
+    if expression.f_op.kind_name in MATH_OPERATIONS:
+        return MATH_OPERATIONS[expression.f_op.kind_name](
+            create_math_expression(expression.f_left, filename, package),
+            create_math_expression(expression.f_right, filename, package),
+            location=node_location(expression, filename),
+        )
+    if expression.f_op.kind_name in MATH_COMPARISONS:
+        return MATH_COMPARISONS[expression.f_op.kind_name](
+            create_math_expression(expression.f_left, filename, package),
+            create_math_expression(expression.f_right, filename, package),
+            location=node_location(expression, filename),
+        )
```

Comment:
I think this part could be shorter if `create_math_expression` and `node_location` were only called once at the beginning.

Reply:
Simplified in 230b019c